### PR TITLE
Comprehensive table operations IT

### LINKED ISSUE #5427: Add test of all table operations against system and user tables
**Is your feature request related to a problem? Please describe.**
Need to ensure all table operations are tested on Accumulo system tables and on user tables

Table operations operate on Accumulo tables including both user tables and system tables. Different operations work differently based on what table they are operating on. For example:
* Operating on user tables, the FATE table, and the SCAN_REF table will operate on that table and potentially metadata in the METADATA table. Operating on the METADATA table will operate on the METADATA table and potentially metadata in the ROOT table. Operating on the ROOT table will operate on the ROOT table and potentially metadata in ZooKeeper.
* In 4.0, table operations that initiate a FATE operation are [stored differently and use different resources](https://github.com/apache/accumulo/blob/f65e15d954669fe3419ec403a4659cd44e926478/server/manager/src/main/java/org/apache/accumulo/manager/Manager.java#L1352-L1355) based on what table they operate on (i.e., USER (for user tables) vs META (for system tables) FATE ops)
* Some ops aren't acceptable on some or all Accumulo system tables. For example, we wouldn't want a user to delete any of the system tables.

Therefore, we should be testing all table operations against all system tables and user tables to ensure they are and always will be functioning as expected.

**Describe the solution you'd like**
Survey the tests to determine if:
There is a test covering all table operations against system and user tables
* If there is, no work needs to be done
* If there is not, add a full-coverage test (or expand on existing test (perhaps `TableOperationsIT`?))

**Additional context**
https://github.com/apache/accumulo/pull/5301 - discussion leading to this issue creation

### PR DESCRIPTION
**Pull request by kevinrr888**
This PR:

Creates a comprehensive IT of all table operations against both system and user tables while trying to avoid duplicating existing testing. For example, when there was more extensive testing of a table operation elsewhere, these new tests just ensure that that testing still exists. This IT does not test for edge cases, but rather tests for basic expected functionality of all table operations against user tables and all system tables (ROOT, METADATA, FATE, SCAN_REF).

Motivation:
- No single test testing all table operations making it hard to know what is or isn't tested
- Lots of operations that aren't tested on all system tables and some operations that are completely untested. Good to ensure each table op works as expected on the different system tables as well as user tables. For example, some ops can't be performed on the metadata tables (ROOT, METADATA) but can be performed on FATE and SCAN_REF, some ops aren't allowed on any system tables, etc.
- Has made identifying potential issues related to what ops are allowed on what system tables easier. For example, have discovered that `offline`ing the FATE and SCAN_REF tables are allowed through the table operations API when they shouldn't be (#5433)
- Ensures end-to-end functionality of table operations always work as expected
- Ensures FATE, as it's used in the Manager, is functioning as expected for both META (FATE ops against system tables) and USER (FATE ops against user tables) FATE ops. This is important since they use different resources and different stores. Lots of individual FATE tests, but this verifies FATE is working as expected in the Manager.
- Will force any newly added table operations to have testing here.
- Will force any new system tables to be considered in these tests.

closes #5427

### CODE REVIEW

File: test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java [hunk 1]
```diff
@@ -0,0 +1,957 @@
...
+          }
+          return numFilesAfterFlush > finalNumFilesBeforeFlush;
+        }
+      });
+    }
+  }
+
+  @Test
+  public void
+      test_setProperty_modifyProperties_removeProperty_getProperties_getTableProperties_getConfiguration()
```
Comment (kevinrr888) [inline test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java:572]: Should setting properties on the system tables continue to be allowed or is this undesired?

File: test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java [hunk 2]
```diff
@@ -0,0 +1,957 @@
...
+        case SCAN_REF:
+          ops.offline(sysTable.tableName(), true);
+          try {
+            assertThrows(AccumuloException.class,
+                () -> ops.exportTable(sysTable.tableName(), exportDir.toString()));
+            assertThrows(AccumuloException.class,
+                () -> ops.importTable(sysTable.tableName(), exportUserDir.toString()));
+          } finally {
+            ops.online(sysTable.tableName(), true);
+          }
```
Comment (kevinrr888) [inline test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java:255]: This will be changed once #5433 is completed and merged in: should not be able to offline any system table

File: test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java [hunk 3]
```diff
@@ -0,0 +1,957 @@
...
+    for (var sysTable : AccumuloTable.values()) {
+      switch (sysTable) {
+        case METADATA:
+        case ROOT:
+          // should not be able to delete rows of METADATA or ROOT tables
+          assertThrows(AccumuloException.class,
+              () -> ops.deleteRows(sysTable.tableName(), null, null));
+          break;
+        case FATE:
+          break;
```
Comment (kevinrr888) [inline test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java:352]: deleteRows() is allowed for FATE table but probably shouldn't be. Can someone confirm that this is an action that should not be allowed? Not testing for now since deleting our FATE operation from the FATE table when it is in progress will cause problems in the MAC and there is no way to put junk data in the FATE table with the Manager running.

File: test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java [hunk 4]
```diff
@@ -0,0 +1,957 @@
...
+        case SCAN_REF:
+          cleanupScanRefTable();
+          // check max row again
+          assertNull(
+              ops.getMaxRow(sysTable.tableName(), Authorizations.EMPTY, null, true, null, true));
+          break;
```
Comment (kevinrr888) [inline test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java:358]: deleteRows() is allowed for SCAN_REF table but probably shouldn't be. Can someone confirm that this is an action that should not be allowed? Testing it here is fine for now since just some junk data in the SCAN_REF table anyways but should probably result in an Exception like METADATA and ROOT.

File: test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java [hunk 5]
```diff
@@ -0,0 +1,957 @@
...
+  public void test_importDirectory() throws Exception {
+    // importDirectory for user tables is tested in BulkNewIT. Ensure test exists
+    assertDoesNotThrow(() -> Class.forName(BulkNewIT.class.getName()));
+    // importDirectory for system tables not tested for. Test basic functionality here
+    var rootPath = getCluster().getTemporaryPath().toString();
+    var dir = rootPath + "/" + getUniqueNames(1)[0];
+    BulkNewIT.writeData(dir + "/f1.", getCluster().getServerContext().getConfiguration(), 0, 5, 0);
+
+    for (var sysTable : AccumuloTable.values()) {
+      assertThrows(Exception.class, () -> ops.importDirectory(dir).to(sysTable.tableName()).load());
```
Comment (kevinrr888) [inline test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java:627]: Some throw AccumuloException and some throw AccumuloSecurityException not sure if that matters. The AccumuloException occurs from FateServiceHandler and the AccumuloSecurityException occurs from ClientServiceHandler. The validation done in FateServiceHandler only checks that it's not the ROOT table. Should probably check that it's not a built in table. This would result in the same Exception for all system tables and is more clear.

File: test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java [hunk 6]
```diff
@@ -0,0 +1,957 @@
...
+          client.tableOperations().online(sysTable.tableName(), true);
+          break;
+        case SCAN_REF:
+        case FATE:
+          assertTrue(client.tableOperations().isOnline(sysTable.tableName()));
+          client.tableOperations().offline(sysTable.tableName(), true);
+          assertFalse(client.tableOperations().isOnline(sysTable.tableName()));
+          client.tableOperations().online(sysTable.tableName(), true);
+          assertTrue(client.tableOperations().isOnline(sysTable.tableName()));
+          break;
```
Comment (kevinrr888) [inline test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java:653]: This will be changed once #5433 is completed and merged in: should not be able to offline any system table

File: test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java [hunk 7]
```diff
@@ -0,0 +1,957 @@
...
+    userTable = getUniqueNames(1)[0];
+    ops.create(userTable);
+    String iterName = "iter_name";
+    var iterSetting = new IteratorSetting(100, iterName, NoOpFilter.class);
+    var scope = EnumSet.of(IteratorUtil.IteratorScope.majc, IteratorUtil.IteratorScope.minc);
+    // system and user tables
+    Set<String> allTables = ops.list();
+
+    for (String table : allTables) {
+      ops.attachIterator(table, iterSetting, scope);
```
Comment (kevinrr888) [inline test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java:699]: Should custom iterators on the system tables continue to be allowed or is this undesired?

File: test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java [hunk 8]
```diff
@@ -0,0 +1,957 @@
...
+    // addConstraint, listConstraints, removeConstraint for user tables is tested in ConstraintIT.
+    // Ensure test exists
+    assertDoesNotThrow(() -> Class.forName(ConstraintIT.class.getName()));
+    // addConstraint, listConstraints, removeConstraint not tested for system tables. Test basic
+    // functionality here
+    for (var sysTable : AccumuloTable.values()) {
+      var numExistingConstraints = ops.listConstraints(sysTable.tableName()).size();
+      String constraint = ComprehensiveBaseIT.TestConstraint.class.getName();
+
+      var constraintNum = ops.addConstraint(sysTable.tableName(), constraint);
```
Comment (kevinrr888) [inline test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java:726]: Should adding constraints on the system tables continue to be allowed or is this undesired?

File: test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java [hunk 9]
```diff
@@ -0,0 +1,957 @@
...
+    createScanRefTableRow();
+    for (var sysTable : AccumuloTable.values()) {
+      ops.flush(sysTable.tableName(), null, null, true);
+      var diskUsageList = ops.getDiskUsage(Set.of(sysTable.tableName()));
+      assertEquals(1, diskUsageList.size());
+      var diskUsage = diskUsageList.get(0);
+      log.info("table : {}, disk usage : {}", sysTable.tableName(), diskUsage.getUsage());
+      if (!sysTable.equals(AccumuloTable.ROOT)) {
+        assertTrue(diskUsage.getUsage() > 0);
+      }
```
Comment (kevinrr888) [inline test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java:757]: Skipping ROOT for now until I resolve the bug in TableDiskUsage.getDiskUsage... Will always show a disk usage of 0 for the ROOT table as it is currently scanning the METADATA table for the ROOT table metadata.

File: test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java [hunk 10]
```diff
@@ -0,0 +1,957 @@
...
+
+  @Test
+  public void test_setTabletAvailability_getTabletInformation() throws Exception {
+    // these are tested for user tables in TabletAvailabilityIT. Ensure test exists
+    assertDoesNotThrow(() -> Class.forName(TabletAvailabilityIT.class.getName()));
+    // these are not tested for system tables. Test basic functionality here
+    for (var sysTable : AccumuloTable.values()) {
+      // should not be able to unhost any system table
+      assertThrows(IllegalArgumentException.class, () -> ops
+          .setTabletAvailability(sysTable.tableName(), new Range(), TabletAvailability.UNHOSTED));
```
Comment (kevinrr888) [inline test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java:830]: Should maybe be an AccumuloException to keep consistency with the other table operations methods that throw an AccumuloException when performed on an unacceptable table?

File: test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java [hunk 11]
```diff
@@ -0,0 +1,957 @@
...
+      }
+    });
+  }
+
+  /**
+   * Cleans up the data in the SCAN_REF table that was created by calls to
+   * {@link #createScanRefTableRow()}
+   */
+  private void cleanupScanRefTable() throws Exception {
+    ops.deleteRows(AccumuloTable.SCAN_REF.tableName(), null, null);
```
Comment (kevinrr888) [inline test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java:920]: Same comment about deleteRows() here, but using it for now since it's helpful cleanup in this test

File: core/src/main/java/org/apache/accumulo/core/clientImpl/ClientTabletCacheImpl.java [hunk 12]
```diff
@@ -660,6 +660,7 @@ private void requestTabletHosting(ClientContext context,
     }
 
     // System tables should always be hosted
+    // TODO should this return for all sys tables?
     if (AccumuloTable.ROOT.tableId() == tableId || AccumuloTable.METADATA.tableId() == tableId) {
       return;
     }
```
Comment (kevinrr888) [inline core/src/main/java/org/apache/accumulo/core/clientImpl/ClientTabletCacheImpl.java:666]: Commenting here as a reminder to delete this once resolved

File: test/src/main/java/org/apache/accumulo/test/functional/CloneTestIT.java [hunk 13]
```diff
@@ -352,18 +352,18 @@ public void testCloneWithSplits() throws Exception {
...
-      assertThrows(AccumuloException.class, () -> client.tableOperations()
-          .clone(AccumuloTable.METADATA.tableName(), "mc1", CloneConfiguration.empty()));
+      var sysTables = AccumuloTable.values();
+      var tableNames = getUniqueNames(sysTables.length);
+
+      for (int i = 0; i < sysTables.length; i++) {
+        var sysTable = sysTables[i];
+        var cloneTableName = tableNames[i];
+        assertThrows(Exception.class, () -> client.tableOperations().clone(sysTable.tableName(),
+            cloneTableName, CloneConfiguration.empty()));
```
Comment (kevinrr888) [inline test/src/main/java/org/apache/accumulo/test/functional/CloneTestIT.java:364]: Same comment about some throwing AccumuloException and some throwing AccumuloSecurityException

File: test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java (hunk 3 above)
Comment (keith-turner) [reply test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java:352]: Seems like it would be ok to add a hard coded check to FateServiceHandler for this like the root and metadata table.

Kinda related, wondering if we have any test that check the default out of the box permissions of the system tables.  Looking into this question I saw that calling this operation requires the write permission on a tablet.  What this operation does can be done much slower w/ a batch writer.

**Review by keith-turner — COMMENTED** [review]
Still looking at this.  Posting comments I have so far.

File: core/src/main/java/org/apache/accumulo/core/clientImpl/ClientTabletCacheImpl.java (hunk 12 above)
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/clientImpl/ClientTabletCacheImpl.java:638]: Created PR #5486

Comment (kevinrr888) [reply core/src/main/java/org/apache/accumulo/core/clientImpl/ClientTabletCacheImpl.java:638]: Created https://github.com/apache/accumulo/pull/5487

File: test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java [hunk 14]
```diff
@@ -0,0 +1,958 @@
...
+    // need to test both
+
+    // addSplits, listSplits tested elsewhere for METADATA, ROOT, and user tables, but testing here
+    // as well since this setup is needed to test for splitRangeByTablets anyway, which is untested
+    // elsewhere
+
+    userTable = getUniqueNames(1)[0];
+    ops.create(userTable);
+    // system and user tables
+    var allTables = new ArrayList<>(ops.list());
```
[the remark this answers is not in the document]
Comment (kevinrr888) [reply test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java]: Since this is already tested in `test_list_tableIdMap`, chose to avoid same checks elsewhere

File: test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java [hunk 15]
```diff
@@ -0,0 +1,958 @@
...
+          break;
+        default:
+          throw new IllegalStateException("Unrecognized table: " + sysTable);
+      }
+    }
+  }
+
+  @Test
+  public void test_addSplits_putSplits_listSplits_splitRangeByTablets() throws Exception {
+    // note that addSplits and putSplits are implemented the same, just take different args. No
```
[the remark this answers is not in the document]
Comment (kevinrr888) [reply test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java]: Yes, good point. Added testing for `putSplits` in 4eb99ebd482139c834dbad28948a1842ab88b7d5

File: test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java [hunk 16]
```diff
@@ -0,0 +1,958 @@
...
+  public void test_list_tableIdMap() throws Exception {
+    // Thoroughly tested elsewhere, but simple enough to test here. Test that all system tables
+    // and user tables are returned
+    userTable = getUniqueNames(1)[0];
+    ops.create(userTable);
+    var expected = Arrays.stream(AccumuloTable.values()).map(AccumuloTable::tableName)
+        .collect(Collectors.toSet());
+    expected.add(userTable);
+    assertEquals(expected, ops.list());
+    assertEquals(expected, ops.tableIdMap().keySet());
```
[the remark this answers is not in the document]
Comment (kevinrr888) [reply test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java:187]: Good point, added testing for values as well in 4eb99ebd482139c834dbad28948a1842ab88b7d5

File: test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java (hunk 2 above)
Comment (kevinrr888) [reply test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java]: #5433 has been merged in. This test has been updated (in 4eb99ebd482139c834dbad28948a1842ab88b7d5) to expect an exception when offlining any system table

File: test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java (hunk 6 above)
Comment (kevinrr888) [reply test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java]: #5433 has been merged in. This test has been updated (in 4eb99ebd482139c834dbad28948a1842ab88b7d5) to expect an exception when offlining any system table

File: test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java (hunk 9 above)
Comment (kevinrr888) [reply test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java]: https://github.com/apache/accumulo/pull/5479 fixed this issue and 4eb99ebd482139c834dbad28948a1842ab88b7d5 removed this skip

**Review by DomGarguilo — APPROVED** [review]

File: test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java [hunk 17]
```diff
@@ -0,0 +1,957 @@
...
+          mut.put("cf", "cq", "v");
+          bw.addMutation(mut);
+        }
+        ops.flush(userTable, null, null, true);
+        ops.flush(AccumuloTable.METADATA.tableName(), null, null, true);
+        ops.flush(AccumuloTable.ROOT.tableName(), null, null, true);
+      }
+
+      for (var sysTable : List.of(AccumuloTable.ROOT, AccumuloTable.METADATA,
+          AccumuloTable.SCAN_REF, AccumuloTable.FATE)) {
```
[the remark this answers is not in the document]
Comment (kevinrr888) [reply test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java:427]: Yeah, it will pass on its own but not when all tests are run because some other tests call `createFateTableRow(userTable)` which is the root cause of the problem.

File: test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java [hunk 18]
```diff
@@ -0,0 +1,958 @@
...
+  }
+
+  @AfterEach
+  public void afterEach() throws Exception {
+    // ensure none of the FATE or SCAN_REF data we created persists between tests. Also ensure the
+    // user table does not persist between tests
+    if (userTable != null && ops.exists(userTable)) {
+      cleanupFateTable();
+      ops.delete(userTable);
+    }
```
[the remark this answers is not in the document]
Comment (kevinrr888) [reply test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java]: Thanks, yeah, should be setting userTable to null here to prevent accidentally reusing the same userTable across tests as well

File: test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java [hunk 19]
```diff
@@ -0,0 +1,958 @@
...
+  }
+
+  @Test
+  public void testAllTested() {
+    var allTableOps = Arrays.stream(TableOperations.class.getDeclaredMethods()).map(Method::getName)
+        .collect(Collectors.toSet());
+    var testMethodNames = Arrays.stream(ComprehensiveTableOperationsIT.class.getDeclaredMethods())
+        .map(Method::getName).collect(Collectors.toSet());
+    var allTableOpsIter = allTableOps.iterator();
+    while (allTableOpsIter.hasNext()) {
```
[the remark this answers is not in the document]
Comment (kevinrr888) [reply test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java]: Certainly much more concise. Added

File: core/src/main/java/org/apache/accumulo/core/clientImpl/ClientTabletCacheImpl.java (hunk 12 above)
Comment (kevinrr888) [reply core/src/main/java/org/apache/accumulo/core/clientImpl/ClientTabletCacheImpl.java]: Resolved

File: test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java (hunk 3 above)
Comment (kevinrr888) [reply test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java:350]: Created https://github.com/apache/accumulo/pull/5491

When testing this, noticed you can call `client.tableOperations.deleteRows` with no problems for FATE and SCAN_REF, but trying to use a batch writer 
(`BatchWriter bw = client.createBatchWriter(AccumuloTable.SCAN_REF.tableName())` for example) you cannot write mutations this way:
`org.apache.accumulo.core.client.MutationsRejectedException: constraint violation codes : []  security codes: {accumulo.scanref(ID:+scanref)=[PERMISSION_DENIED]}  # server errors 0 # exceptions 0`

File: test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java (hunk 11 above)
Comment (kevinrr888) [reply test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java:900]: Created https://github.com/apache/accumulo/pull/5491

File: test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java (hunk 7 above)
Comment (keith-turner) [reply test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java:681]: That could be dangerous, but I think its fine to leave it because there are many other dangerous things that could potentially be done.   Overall the best option is to ensure the metadata table has locked down permissions for setting table properties and writing to it.

File: test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (keith-turner) [reply test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java:421]: I can take a look into this. If this is not figured out before this PR is merged then we should open an issue for this.

File: test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java (hunk 5 above)
Comment (keith-turner) [reply test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java:625]: For these different exceptions you are seeing, do they seem correct?  Like is the AccumuloException related to a not security issue and it the AccumuloSecurityException related to a security issue?

File: test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java [hunk 20]
```diff
@@ -0,0 +1,937 @@
...
+  }
+
+  @Test
+  public void test_setLocalityGroups_getLocalityGroups() throws Exception {
+    // setLocalityGroups, getLocalityGroups for user tables is tested in LocalityGroupIT. Ensure
+    // test exists
+    assertDoesNotThrow(() -> Class.forName(LocalityGroupIT.class.getName()));
+    // setLocalityGroups, getLocalityGroups for system tables not tested for. Test basic
+    // functionality here
+    for (var sysTable : AccumuloTable.values()) {
```
Comment (keith-turner) [inline test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java:608]: Looking into this test noticed #5494

File: test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java [hunk 21]
```diff
@@ -0,0 +1,937 @@
...
+    userTable = getUniqueNames(1)[0];
+    ops.create(userTable);
+    // system and user tables
+    Set<String> allTables = ops.list();
+    for (String tableName : allTables) {
+      // setProperty
+      String propKey = Property.TABLE_ARBITRARY_PROP_PREFIX.getKey() + "prop";
+      String propVal = "val";
+      String newPropVal = "newval";
+      ops.setProperty(tableName, propKey, propVal);
```
Comment (keith-turner) [inline test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java:581]: Before initially setting the prop could add a check to ensure its not already there.

File: test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java [hunk 22]
```diff
@@ -0,0 +1,937 @@
...
+      // Wait until we see the new file
+      final int finalNumFilesBeforeFlush = numFilesBeforeFlush;
+      Wait.waitFor(() -> {
+        try (TabletsMetadata tabletsMetadata = getCluster().getServerContext().getAmple()
+            .readTablets().forTable(sysTable.tableId()).build()) {
+          int numFilesAfterFlush = 0;
+          for (var tm : tabletsMetadata) {
+            numFilesAfterFlush += tm.getFiles().size();
+          }
+          return numFilesAfterFlush > finalNumFilesBeforeFlush;
```
Comment (keith-turner) [inline test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java:561]: comparing counts could cause failures because these tables are configured to compact aggresively.  So if the table compacts after the flush and before counting is done in the test it could throw things off.  Instead of comparing counts, could compare that the set of files changed.

File: test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (kevinrr888) [reply test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java:421]: That would be great.

File: test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java (hunk 11 above)
Comment (kevinrr888) [reply test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java]: Replaced deleteRows here. deleteRows on system tables is no longer possible

File: test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java (hunk 3 above)
Comment (kevinrr888) [reply test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java]: deleteRows on system tables is no longer possible. Test now expects deleteRows to throw an exception.

File: test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java (hunk 4 above)
Comment (kevinrr888) [reply test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java]: deleteRows on system tables is no longer possible. Test now expects deleteRows to throw an exception.

File: test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java (hunk 22 above)
Comment (kevinrr888) [reply test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java]: Good point. Addressed in c88263153459d77dad6ce76bb3c7ac4449bdbde1

File: test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java (hunk 21 above)
Comment (kevinrr888) [reply test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java:570]: Added in c88263153459d77dad6ce76bb3c7ac4449bdbde1

File: test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java (hunk 1 above)
Comment (kevinrr888) [reply test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java:553]: Verified that permissions are checked

File: test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java (hunk 5 above)
Comment (kevinrr888) [reply test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java:619]: With BULK_IMPORT2, the table is validated:
`final var tableId = validateTableIdArgument(arguments.get(0), tableOp, NOT_ROOT_TABLE_ID);`
Allowing tableIds that aren't the ROOT table to pass this check.
Then, the table is further validated to check if the user has permission to BULK_IMPORT2 into the table:
`canBulkImport = security.canBulkImport(c, tableId, tableName, dir, null, namespaceId);`

I think the validation of the tableId should include all system tables. I don't think we ever would want a user to bulk import to a system table, regardless of permissions.

File: test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java (hunk 10 above)
Comment (kevinrr888) [reply test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java:804]: Opened https://github.com/apache/accumulo/pull/5510

A lot of the table operations don't have javadoc for what the thrown exceptions mean, so maybe this can be it's own ticket

File: test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (kevinrr888) [reply test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java]: Opened https://github.com/apache/accumulo/issues/5679

File: test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java (hunk 10 above)
Comment (kevinrr888) [reply test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java]: Functionality fixed in #5510, assertion fixed in 937e4f8c595b128e94a32ab0e2b687a0ff8cfe31

File: test/src/main/java/org/apache/accumulo/test/functional/CloneTestIT.java (hunk 13 above)
Comment (kevinrr888) [reply test/src/main/java/org/apache/accumulo/test/functional/CloneTestIT.java]: Created #5680

File: test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java (hunk 5 above)
Comment (kevinrr888) [reply test/src/main/java/org/apache/accumulo/test/ComprehensiveTableOperationsIT.java]: Created #5680
